pkg/query-service/app/logs/v3/json_filter_test.go

```diff
@@ -0,0 +1,250 @@
+	{
+		Name: "Array Float",
+		Key: v3.AttributeKey{
+			Key:      "body.nested_num[*].float_nums[*]",
+			DataType: "array(float64)",
+			IsJSON:   true,
+		},
+		ClickhouseKey: "JSONExtract(JSON_QUERY(body, '$.nested_num[*].float_nums[*]'), '" + ARRAY_FLOAT64 + "')",
+	},
+	{
+		Name: "String",
+		Key: v3.AttributeKey{
+			Key:      "body.message",
+			DataType: "string",
+			IsJSON:   true,
+		},
+		ClickhouseKey: "JSON_VALUE(body, '$.message')",
+	},
+	{
+		Name: "Int",
+		Key: v3.AttributeKey{
+			Key:      "body.status",
+			DataType: "int64",
+			IsJSON:   true,
+		},
+		ClickhouseKey: "JSONExtract(JSON_QUERY(body, '$.status'), '" + INT64 + "')",
+	},
+	{
+		Name: "Float",
+		Key: v3.AttributeKey{
+			Key:      "body.fraction",
+			DataType: "float64",
+			IsJSON:   true,
+		},
+		ClickhouseKey: "JSONExtract(JSON_QUERY(body, '$.fraction'), '" + FLOAT64 + "')",
+	},
+}
+
+func TestGetJSONFilterKey(t *testing.T) {
+	for _, tt := range testGetJSONFilterKeyData {
+		Convey("testgetKey", t, func() {
+			columnName, err := getJSONFilterKey(tt.Key)
+			if tt.Error {
+				So(err, ShouldNotBeNil)
+			} else {
+				So(err, ShouldBeNil)
+				So(columnName, ShouldEqual, tt.ClickhouseKey)
+			}
+		})
+	}
+}
+
+var testGetJSONFilterData = []struct {
+	Name       string
+	FilterItem v3.FilterItem
+	Filter     string
+	Error      bool
+}{
+	{
+		Name: "Array membership string",
+		FilterItem: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.requestor_list[*]",
+				DataType: "array(string)",
+				IsJSON:   true,
+			},
+			Operator: "has",
+			Value:    "index_service",
+		},
+		Filter: "has(JSONExtract(JSON_QUERY(body, '$.requestor_list[*]'), 'Array(String)'), 'index_service')",
+	},
+	{
+		Name: "Array membership int64",
+		FilterItem: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.int_numbers[*]",
+				DataType: "array(int64)",
+				IsJSON:   true,
+			},
+			Operator: "has",
+			Value:    2,
+		},
+		Filter: "has(JSONExtract(JSON_QUERY(body, '$.int_numbers[*]'), '" + ARRAY_INT64 + "'), 2)",
+	},
+	{
+		Name: "Array membership float64",
+		FilterItem: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.nested_num[*].float_nums[*]",
+				DataType: "array(float64)",
+				IsJSON:   true,
+			},
+			Operator: "has",
+			Value:    2.2,
+		},
+		Filter: "has(JSONExtract(JSON_QUERY(body, '$.nested_num[*].float_nums[*]'), '" + ARRAY_FLOAT64 + "'), 2.200000)",
+	},
+	{
+		Name: "eq operator",
+		FilterItem: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.message",
+				DataType: "string",
+				IsJSON:   true,
+			},
+			Operator: "=",
+			Value:    "hello",
+		},
+		Filter: "JSON_VALUE(body, '$.message') = 'hello'",
+	},
+	{
+		Name: "eq operator number",
+		FilterItem: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.status",
+				DataType: "int64",
+				IsJSON:   true,
+			},
+			Operator: "=",
+			Value:    1,
+		},
+		Filter: "JSONExtract(JSON_QUERY(body, '$.status'), '" + INT64 + "') = 1",
```

Comment:
JSONExtract returns a JSON Object (an array in this case), so comparison with scalar value results in an Exception. Please run all these resulting filter expressions in `[json_filter_test.go]` against the ClickHouse and check.

Reply:
So all of these are tested against clickhouse with actual data.

Reply:
Test the correctness

Reply:
It returns the dataType you specify https://clickhouse.com/docs/en/sql-reference/functions/json-functions#jsonextractjson-indices_or_keys-return_type

Reply:
```
SELECT JSON_QUERY('{"status":200}', '$.status')

Query id: b2773df7-1709-44cc-955d-0dd6108c870e

┌─JSON_QUERY('{"status":200}', '$.status')─┐
│ [200]                                    │
└──────────────────────────────────────────┘
```

```
SELECT JSONExtract(JSON_QUERY('{"status":200}', '$.status'), 'Int64')

Query id: 1361c7e8-433d-47a2-a63f-dc3d9f5ebcec

┌─JSONExtract(JSON_QUERY('{"status":200}', '$.status'), 'Int64')─┐
│                                                              0 │
└────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.004 sec.
```

```
SELECT JSONExtract(JSON_QUERY('{"status":201}', '$.status'), 'Int64') > 200

Query id: f416e77e-5515-42f5-b568-c07e5a783274

┌─greater(JSONExtract(JSON_QUERY('{"status":201}', '$.status'), 'Int64'), 200)─┐
│                                                                            0 │
└──────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.003 sec.
```

Reply:
~JSONExtract~ JSON_QUERY (I meant to say in the first comment) returns a JSON Object (an array in this case)

Reply:
Got the issue, should have use JSON value instead of JSON Query.